### PR TITLE
dup: add duplication_sync_timer

### PR DESCRIPTION
**Pull request by neverchanje**
This is the first part of the implementation of duplication on the replica-server side.

## Introduction

`duplication_sync_timer` is the timer periodically (10 seconds) sends `duplication_sync` to meta-server for retrieving the entire set of duplications on this server (aka `replica_stub`).

```
         +-----------------------+
         |                       |
         |     meta server       |
         |                       |
         +-------^--------+------+
                 |        |
duplication_sync_request  duplication_sync_response
                 |        |
                 |        |
             +------------+---+
             |   |            |
         +---+---+--------v---+--+
         | duplication_sync_timer|
         |   |                |  |
         |   +----------------+  |
         |                       |
         |    replica server     |
         |                       |
         +-----------------------+
```

If the replica server successfully get the response from meta-server (`on_duplication_sync_reply`), 
it will update its local duplication map (`duplication_sync_timer::update_duplication_map`) to:

1. assign new duplications
2. remove unused duplications
3. synchronize the `confirmed_decree` of each primary replica that's on duplication.

Each of the replica (which may be secondary) has one `replica_duplicator_manager` managing the entire set of duplications that it is assigned (each duplicating destination is tasked by one duplication)

```
+-----------------------------------+
|                                   |
|                                   |
|        +----replica-------+       |
|        |                  |       |
|        +------------------+       |
|                                   |
|     +-------replica---------+     |
|     |                       |     |
|  +--+-----------------------+--+  |
|  | replica_duplicator_manager  |  |
|  +--+-----------------------+--+  |
|     |                       |     |
|     +-----------------------+     |
|                                   |
|        +---+replica+------+       |
|        |                  |       |
|        +------------------+       |
|                                   |
+-----------------------------------+
```

### CODE REVIEW

**Comment by hycdong** [comment]
Could you please explain what this pr is doing?

**Comment by hycdong** [comment]
I try to left a comment, but wrongly click button "Close and comment" rather than "Comment". Sorry for my misoperation.

**Comment by hycdong** [comment]
Could you please add some explanation about this pr?

**Comment by neverchanje** [comment]
@hycdong added.

File: src/dist/replication/lib/replica_stub.cpp [hunk 1]
```diff
@@ -37,6 +37,7 @@
 #include "replica_stub.h"
 #include "mutation_log.h"
 #include "mutation.h"
+#include "duplication/duplication_sync_timer.h"
```
Comment (vagetablechicken) [inline src/dist/replication/lib/replica_stub.cpp]: 为什么要在cpp里才include timer？头文件里直接include，可以免去声明吧？

Comment (neverchanje) [reply src/dist/replication/lib/replica_stub.cpp]: 这是 c++ best practices，在 replica_stub.h 头文件 include，那 duplication_sync_timer.h 每次修改，使用了  replica_stub.h 的 .cpp 代码都要重新编译，这个问题又叫 **头文件污染**。所以建议头文件只有声明，不 include，参考 https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/WhyIWYU.md#why-forward-declare。

**Comment by neverchanje** [comment]
@vagetablechicken `duplication_sync_timer_test` 依赖其他还没有合入的代码，暂时不放在这个 PR

**Comment by vagetablechicken** [comment]
> @vagetablechicken `duplication_sync_timer_test` 依赖其他还没有合入的代码，暂时不放在这个 PR

OK

File: src/dist/replication/lib/replica_stub.cpp (hunk 1 above)
Comment (vagetablechicken) [reply src/dist/replication/lib/replica_stub.cpp]: 嗯，明白。但是前置声明的缺点也不少，挺难判定什么时候使用什么时候直接include头文件。那就按这个风格来吧，这里的情况也不复杂。

File: src/dist/replication/lib/duplication/duplication_sync_timer.cpp [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        }
+    }
+
+    if (_timer_task) {
+        _timer_task->cancel(true);
+        _timer_task = nullptr;
+    }
+}
+
+void duplication_sync_timer::start()
```
Comment (hycdong) [inline src/dist/replication/lib/duplication/duplication_sync_timer.cpp]: When this 'start' in called?

File: src/dist/replication/lib/duplication/duplication_sync_timer.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    /// \param dup_map: <appid -> list<dup_entry>>
+    void
+    update_duplication_map(const std::map<app_id, std::map<dupid_t, duplication_entry>> &dup_map);
+
+    void on_duplication_sync_reply(error_code err, const duplication_sync_response &resp);
+
+    std::vector<replica_ptr> get_all_primaries();
+
+    std::vector<replica_ptr> get_all_replicas();
```
Comment (hycdong) [inline src/dist/replication/lib/duplication/duplication_sync_timer.h]: What the usage of 'get_all_primaries' and 'get_all_ replicas'?

Comment (neverchanje) [reply src/dist/replication/lib/duplication/duplication_sync_timer.h]: - `get_all_ replicas`: see `duplication_sync_timer::update_duplication_map`
- `get_all_primaries`: see `duplication_sync_timer::run`

File: src/dist/replication/lib/duplication/duplication_sync_timer.cpp (hunk 2 above)
Comment (neverchanje) [reply src/dist/replication/lib/duplication/duplication_sync_timer.cpp]: https://github.com/XiaoMi/rdsn/pull/281/commits/371947ec89d521f125b2478f208b0ea1ff3b283c

See `replica_stub::initialize_start`, I have committed the code point.

**Pull request merged**